analyzer/runtime/extract.go

```diff
@@ -578,6 +586,84 @@ func sumGasUsed(events []*EventData) (sum uint64, foundGasUsedEvent bool) {
 	return
 }
 
+// Attempts to extract the human-readable error message from
+// the raw error returned by the node.
+// If the function successfully extracts the message, it returns (msg, nil).
+// If the function cannot extract the message, it returns (nil, nil).
+// If the function encounters a malformed error message, it returns (nil, err).
+//
+// In Emerald and older Sapphire versions, transaction revert
+// reasons were decoded by the runtime and returned as
+// human-readable strings, e.g.
+// - "reverted: Incorrect premium amount."
+// - "execution failed: out of gas"
+//
+// More recent versions of Sapphire no longer decode the error
+// message and instead use the following error format:
+// - "reverted: " || base64(abiEncode(error))
+//
+// We first check to see if the error message is a plaintext error string,
+// in which case we simply sanitize and return the message.
+// Otherwise, we optimistically try to decode the error as the prevailing
+// Error(string) type and return the error message if successful.
+//
+// Note that the error type can be any data type specified in the abi.
+// We may not have the abi available now, so in those cases the
+// abi analyzer will extract the error message once the contract
+// abi is available.
+func tryParseErrorMessage(msg string) (*string, error) {
+	// Decode revert reason, format is as follows:
+	// "reverted: " || base64(abiEncode(error))
+	ErrRevertedPrefix := "reverted: "
+	if len(msg) < 10 || msg[:10] != ErrRevertedPrefix {
+		// Likely a plaintext error message.
+		sanitizedMsg := sanitizeString(msg)
+		return &sanitizedMsg, nil
+	}
+	abiEncodedErr, err := base64.StdEncoding.DecodeString(msg[10:])
+	if err != nil {
+		// Likely a plaintext error message.
+		//
+		// Note: This is an imperfect heuristic, some older error messages may
+		// be valid b64 encodings. For newer errors, the runtime guarantees
+		// that the message is base64-encoded. However, if this assumption
+		// ever fails, then the full raw error message will be returned.
+		sanitizedMsg := sanitizeString(msg)
+		return &sanitizedMsg, nil
+	}
+	// If the error type is Error(string), the abi encoding format is as follows:
+	//
+	// 08c379a0                                                         <- Function selector
+	// 0000000000000000000000000000000000000000000000000000000000000020 <- Offset of string return value
+	// 0000000000000000000000000000000000000000000000000000000000000047 <- Length of string return value (the revert reason)
+	// 6d7946756e6374696f6e206f6e6c79206163636570747320617267756d656e74 <- First 32 bytes of the revert reason
+	// 7320776869636820617265206772656174686572207468616e206f7220657175 <- Next 32 bytes of the revert reason
+	// 616c20746f203500000000000000000000000000000000000000000000000000 <- Last 7 bytes of the revert reason
+	//
+	ErrorStringSelector, _ := hex.DecodeString("08c379a0") // Keccak256("Error(string)")
+	const FieldOffsetStart int64 = 4
```

Comment:
Does this parsing code roughly equal to:
```
	typ, _ := abi.NewType("string", "", nil)
	unpacked, err := (abi.Arguments{{Type: typ}}).Pack(msg[10:])
```

This is roughly how web3-gateway used to try optimistically parsing the error message.

Reply:
nicee didn't know there was a library for it - updated